src/allocation_engine/README.md

```diff
@@ -0,0 +1,170 @@
+# Design
+
+## Allocation engine
+
+This implementation uses an interval tree that is specialised for allocation of
+memory-mapped I/O and port I/O address space. The fields of the strctures
+defined will have semantic meaning for this context (e.g node state to indicate
+if a node in the tree is assigned or not to a device).
+
+We offer three options for placing a memory slot in the managed address space:
+
+1. `LastMatch` -> When using this allocation policy the allocator will try to
+insert the range described by the constraint at the first available position
+starting from the end of the managed address space.
+2. `FirstMatch` -> When using this allocation policy the allocator will try to
+insert the range described by the constraint at the first available position
+starting from the begining of the managed address space.
+3. `ExactMatch` -> When using this allocation policy the allocator will try to
+insert the range at the exact position described by the constraint, otherwise
+it will return an error.
+
+```rust
+/// Policy for resource allocation.
+pub enum AllocPolicy {
+    /// Allocate from the first matched entry.
+    FirstMatch,
+    /// Allocate first matched entry from the end of the range.
+    LastMatch,
+    // Allocate a memory slot starting with the specified addrres
+    // if it is available.
+    ExactMatch,
+}
+```
+
+Struct `Constraint` is used to describe the overall information of the resource
+ needed to be allocated. This structure is also used by IntervalTree to know where
+ and how to allocate the resource.
+
+```rust
+/// Struct to describe resource allocation constraints.
+pub struct Constraint {
+    /// Size to allocate.
+    pub size: u64,
+    /// Lower boundary for the allocated resource.
+    pub min: u64,
+    /// Upper boundary for the allocated resource.
+    pub max: u64,
+    /// Alignment for the allocated resource.
+    pub align: u64,
+    /// Resource allocation policy.
+    pub policy: AllocPolicy,
+}
+```
+
+## Interval tree
+
+An interval tree is a tree data structure used for storing information about intervals.
+Specifically, it allows one to efficiently identify intervals that are overlaping
+with a given point, or another interval. We considered that this characteristic
+makes this data structure appropriate to be used as an allocation engine for
+memory slots inside an address space. The time complexity of an interval tree,
+namely O(log ⁡n+m) for  queries, O(log n) for creation and O(log n) for insertion
+and deletion of nodes.
+
+ ```rust
+/// A closed interval range [min, max] used to describe a
+/// memory slot that will be assigned to a device by the VMM.
+/// This structure represents the key of the Node object in
+/// the interval tree implementation.
+pub struct Range {
+    pub min: u64,
+    pub max: u64,
+}
+
+// Node state for interval tree nodes.
+///
+/// Valid state transition:
+/// - None -> Free: IntervalTree::insert()
+/// - Free -> Allocated: IntervalTree::allocate()
```

Comment:
It seems like there's no separate `allocate()` method, but `insert(key, NodeState::Allocated)` instead.